### ISSUE TITLE
Can't build project on latest master with GLFW_BUILD_WAYLAND=ON due to redefinition errors

### ISSUE THREAD

**Issue by warisb** [glfw/glfw#2016]
OS and version:  Ubuntu 18.04
Compiler version: GCC 7.5.0
Release or commit: master, commit 7d060ba4f1237c87ed53e007db619da8c25df282
Build log:
```
wboonyasiriwat@tervdilinaws53:~/dev/glfw-waris/build$ cmake -GNinja -DCMAKE_BUILD_TYPE=Debug -DGLFW_BUILD_WAYLAND=1 -DCMAKE_INSTALL_PREFIX=~/install -DBUILD_SHARED_LIBS=ON ..
-- The C compiler identification is GNU 7.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- Including Wayland support
-- Including X11 support
-- Looking for include file xkbcommon/xkbcommon-compose.h
-- Looking for include file xkbcommon/xkbcommon-compose.h - found
-- Looking for memfd_create
-- Looking for memfd_create - found
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Checking for module 'wayland-protocols>=1.15'
--   Found wayland-protocols, version 1.20
-- Checking for modules 'wayland-client>=0.2.7;wayland-cursor>=0.2.7;wayland-egl>=0.2.7;xkbcommon'
--   Found wayland-client, version 1.16.0
--   Found wayland-cursor, version 1.16.0
--   Found wayland-egl, version 18.1.0
--   Found xkbcommon, version 0.8.2
-- Found X11: /usr/include   
-- Looking for XOpenDisplay in /usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so
-- Looking for XOpenDisplay in /usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/wboonyasiriwat/dev/glfw-waris/build
wboonyasiriwat@tervdilinaws53:~/dev/glfw-waris/build$ 
wboonyasiriwat@tervdilinaws53:~/dev/glfw-waris/build$ ninja
[33/114] Building C object src/CMakeFiles/glfw.dir/wl_init.c.o
FAILED: src/CMakeFiles/glfw.dir/wl_init.c.o 
/usr/bin/cc -DHAVE_MEMFD_CREATE -DHAVE_XKBCOMMON_COMPOSE_H -D_DEFAULT_SOURCE -D_GLFW_BUILD_DLL -D_GLFW_WAYLAND -D_GLFW_X11 -I/home/wboonyasiriwat/dev/glfw-waris/include -I/home/wboonyasiriwat/dev/glfw-waris/src -I/home/wboonyasiriwat/dev/glfw-waris/build/src -g -fPIC -Wall -fvisibility=hidden -std=c99 -MD -MT src/CMakeFiles/glfw.dir/wl_init.c.o -MF src/CMakeFiles/glfw.dir/wl_init.c.o.d -o src/CMakeFiles/glfw.dir/wl_init.c.o -c /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:52:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-xdg-shell-client-protocol-code.h:53:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:51:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-client-protocol-code.h:61:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:53:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-xdg-decoration-client-protocol-code.h:43:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:52:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-xdg-shell-client-protocol-code.h:53:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:54:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-viewporter-client-protocol-code.h:43:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:53:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-xdg-decoration-client-protocol-code.h:43:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:55:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-relative-pointer-unstable-v1-client-protocol-code.h:44:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:54:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-viewporter-client-protocol-code.h:43:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:56:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-pointer-constraints-unstable-v1-client-protocol-code.h:47:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:55:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-relative-pointer-unstable-v1-client-protocol-code.h:44:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:57:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-idle-inhibit-unstable-v1-client-protocol-code.h:43:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from /home/wboonyasiriwat/dev/glfw-waris/src/wl_init.c:56:0:
/home/wboonyasiriwat/dev/glfw-waris/build/src/wayland-pointer-constraints-unstable-v1-client-protocol-code.h:47:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
[50/114] Building C object examples/CMakeFiles/splitview.dir/splitview.c.o
ninja: build stopped: subcommand failed.
wboonyasiriwat@tervdilinaws53:~/dev/glfw-waris/build$ 

```

Note that I have wayland-protocols version 1.15 built and installed from source. I also tried with 1.20 and did not help.

**Comment by elmindreda** [comment]
That doesn't look good. What version of wayland-scanner do you have?

**Comment by warisb** [comment]
@elmindreda seems like 1.16.0

```
$ wayland-scanner --version
wayland-scanner 1.16.0
```

**Comment by warisb** [comment]
I found what seems to be a glaring problems.

1. [These lines](https://github.com/glfw/glfw/blob/master/src/CMakeLists.txt#L86-L91) should produce a source file (.c) not a header. So there's no need to include these files in `wl_init.c`
2. Even after I got around that, I started hitting undefined symbols around [these lines](https://github.com/glfw/glfw/blob/master/src/wl_platform.h#L85-L123). I don't quite know the purpose of these lines but removing these lines don't seem to cause any harm
3. wayland-client [include paths are included](https://github.com/glfw/glfw/blob/master/src/CMakeLists.txt#L174) but its libraries are not linked 

There are some things (especially 2) which I might be missing context for, but I'm going to PR anyway as it seems to be working for me.

PR is up at #2032. Please check.

**Comment by elmindreda** [comment]
> I found what seems to be glaring problems.

These things were done on purpose in dffe203c178ecea44a0c3d52a7b62c45a6615944 to allow the Wayland libraries to be on demand loaded at runtime. This hopefully also addresses point 3.

> 1. [These lines](https://github.com/glfw/glfw/blob/master/src/CMakeLists.txt#L86-L91) should produce a source file (.c) not a header. So there's no need to include these files in `wl_init.c`

They are included as header files so we can redirect the Wayland core function calls in them to our own functions pointers without needing to modify wayland-scanner or its output.

> 2. Even after I got around that, I started hitting undefined symbols around [these lines](https://github.com/glfw/glfw/blob/master/src/wl_platform.h#L85-L123). I don't quite know the purpose of these lines but removing these lines don't seem to cause any harm

Because GLFW is often used as a static library, these macros ensure that our object files don't export symbols that could interfere with other Wayland code during linking.

Since these structs are not static and also grow over time, if the linker ever picks the wrong one when combining GFLW and other Wayland using code compiled with a different version of the protocols, it would result in very difficult to diagnose bugs.

**Comment by elmindreda** [comment]
The slightly hairy approach of 65617b9abaca3d5edb8c52a292d13a8755778a57 in the `wayland-scanner-compatibility` branch seems to work on my copy of Ubuntu 18.04, without losing the benefits of dynamic loading.

**Comment by elmindreda** [comment]
This should be fixed now with 65617b9abaca3d5edb8c52a292d13a8755778a57. Thank you for reporting this!
